internal/querycoordv2/observers/target_observer.go

```diff
@@ -0,0 +1,136 @@
+package observers
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/querycoordv2/meta"
+	"github.com/milvus-io/milvus/internal/querycoordv2/params"
+	"github.com/milvus-io/milvus/internal/querycoordv2/utils"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+)
+
+type TargetObserver struct {
+	c         chan struct{}
+	wg        sync.WaitGroup
+	meta      *meta.Meta
+	targetMgr *meta.TargetManager
+	distMgr   *meta.DistributionManager
+	broker    meta.Broker
+
+	nextTargetLastUpdate map[int64]time.Time
+}
+
+func NewTargetObserver(meta *meta.Meta, targetMgr *meta.TargetManager, distMgr *meta.DistributionManager, broker meta.Broker) *TargetObserver {
+	return &TargetObserver{
+		c:                    make(chan struct{}),
+		meta:                 meta,
+		targetMgr:            targetMgr,
+		distMgr:              distMgr,
+		broker:               broker,
+		nextTargetLastUpdate: make(map[int64]time.Time),
+	}
+}
+
+func (ob *TargetObserver) Start(ctx context.Context) {
+	ob.wg.Add(1)
+	go ob.schedule(ctx)
+}
+
+func (ob *TargetObserver) Stop() {
+	close(ob.c)
+	ob.wg.Wait()
+}
+
+func (ob *TargetObserver) schedule(ctx context.Context) {
+	defer ob.wg.Done()
+	log.Info("Start update next target loop")
+
+	ticker := time.NewTicker(params.Params.QueryCoordCfg.UpdateNextTargetInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Close target observer due to context canceled")
+			return
+		case <-ob.c:
+			log.Info("Close target observer")
+			return
+
+		case <-ticker.C:
+			ob.tryUpdateTarget()
+		}
+	}
+}
+
+func (ob *TargetObserver) tryUpdateTarget() {
+	collections := ob.meta.GetAll()
+	for _, collectionID := range collections {
+		if ob.shouldUpdateCurrentTarget(collectionID) {
+			ob.updateCurrentTarget(collectionID)
+		}
+
+		if ob.shouldUpdateNextTarget(collectionID) {
```

Comment:
Why dont we try to pull next target only when current is completed but pull it intervally?

Reply:
for case when current target's update is not trigger but next target is expired